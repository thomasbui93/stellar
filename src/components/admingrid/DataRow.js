import React from 'react'
import {Link} from 'react-router-dom'

export const DataRow = ({className, item}) => {
  return <div className={className}>
    <div className={`${className}__name`}>{item.name}</div>
    <div className={`${className}__key`}>{item.key}</div>
    <div className={`${className}__status`}>{item.isActive ? 'Active' : 'Archived'}</div>
    <div className={`${className}__action`}>
      <div className={`${className}__remove`} onClick={() => {item.removeAction(item.key)}} >Delete</div>
      <Link className={`${className}__edit`} to={`${item.baseUrl}/${item.key}`}>Edit</Link>
    </div>
  </div>
}
