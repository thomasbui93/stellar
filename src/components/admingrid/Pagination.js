import React from 'react'
import PropTypes from 'prop-types'

const Pagination = ({next, prev, requestApi}) => {
  return (
    <div className='pagination'>
      { next > 0 ? <div className='pagination-next' onClick={() => { requestApi({page: next}) }}>Next</div> : ''}
      { prev > 0 ? <div className='pagination-previous' onClick={() => { requestApi({page: prev}) }}>Prev</div> : ''}
    </div>
  )
}

Pagination.propTypes = {
  next: PropTypes.number,
  prev: PropTypes.number,
  requestApi: PropTypes.func
}

export { Pagination }
