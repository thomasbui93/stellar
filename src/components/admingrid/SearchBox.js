import React, {Component} from 'react'
import PropTypes from 'prop-types'

export class SearchBox extends Component {
  static propTypes = {
    requestApi: PropTypes.func
  };

  constructor(props) {
    super(props);
    this.submitAction = this.submitAction.bind(this);
    this.clearSearch = this.clearSearch.bind(this);
  }

  submitAction(event) {
    event.preventDefault();
    const query = this.refs.query.value;
    if (query.length >= 3) {
      this.props.requestApi({query: query})
    }
  }

  clearSearch(){
    this.refs.query.value = '';
    this.props.requestApi();
  }

  render() {
    return <form className={`search-box ${this.props.className ? this.props.className : ''}`}
                 onSubmit={this.submitAction}>
      <input placeholder="Enter your search here" type="text" ref='query'/>
      <button type="submit">Search</button>
      <button type="button" onClick={this.clearSearch}>
        Clear
      </button>
    </form>
  }
}