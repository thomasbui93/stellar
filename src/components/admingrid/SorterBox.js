import React, {Component} from 'react'
import PropTypes from 'prop-types'

export class SorterBox extends Component {
  constructor (props) {
    super(props)
    this.applySortAction = this.applySortAction.bind(this)
  }

  static propTypes = {
    requestApi: PropTypes.func,
    fields: PropTypes.arrayOf(PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.label
    }))
  };

  applySortAction () {
    this.props.requestApi({sort: this.refs.field.value, desc: this.refs.dir.value})
  }

  render () {
    const {fields} = this.props
    return <div className='sorter-box'>
      {fields.length > 0
        ? <div>
          <select className='sorter-box__field' ref='field' onChange={this.applySortAction}>
            { fields.map(({value, label}) => {
              return <option value={value} key={value}> {label}</option>
            })}
          </select>
          <select className='sorter-box__direction' ref='dir' onChange={this.applySortAction}>
            <option value={'true'} >DESC</option>
            <option value={'false'}>ASC</option>
          </select>
        </div>
        : ''
      }
    </div>
  }
}
