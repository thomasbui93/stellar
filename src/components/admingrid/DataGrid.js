import React from 'react'
import {DataRow} from './DataRow'

export const DataGrid = props => {
  const {items, rowClassName, baseUrl, className, removeAction} = props;
  return <div className={`admin-data-grid ${className}`}>
    {items && items.length > 0 ? renderList(items, rowClassName, baseUrl, removeAction) : <p className="empty-grid">Your grid is currently empty</p>}
  </div>
};

const renderList = (items, rowClassName, baseUrl, removeAction) => {
  return items.map(item => {
    const populatedItem = {...item, baseUrl, removeAction};
    return <DataRow item={populatedItem} className={rowClassName} key={item.key} />
  })
};
