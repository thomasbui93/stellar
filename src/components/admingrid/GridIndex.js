import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {DataGrid} from './DataGrid'
import {StaticTag} from '../general/StaticTag'
import {Loader} from '../../components/general/Loader'
import {Pagination} from './Pagination'
import {SearchBox} from './SearchBox'
import {SorterBox} from './SorterBox'

export class GridIndex extends Component {
  static propTypes = {
    items: PropTypes.array,
    isLoading: PropTypes.bool,
    removeAction: PropTypes.func,
    requestApi: PropTypes.func,
    sortingFields: PropTypes.arrayOf(PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string
    })),
    pagination: PropTypes.shape({
      next: PropTypes.number,
      prev: PropTypes.number
    }),
    error: PropTypes.oneOfType([PropTypes.string, PropTypes.bool])
  };

  componentDidMount () {
    this.props.requestApi()
  }

  render () {
    const {
      error,
      isLoading,
      items,
      removeAction,
      match,
      pagination,
      requestApi,
      sortingFields
    } = this.props

    return <div className={this.props.className}>
      <div className='messages'>
        {error ? <StaticTag className='request-error' staticText={error} /> : ''}
      </div>
      <div className='filter-box'>
        <SearchBox requestApi={requestApi} />
        <SorterBox requestApi={requestApi} fields={sortingFields} />
      </div>
      <Pagination requestApi={requestApi} next={pagination.next} prev={pagination.prev} />
      <div>
        {isLoading ? <Loader /> : ''}
        <DataGrid
          baseUrl={match.url}
          items={items}
          removeAction={removeAction} />
      </div>
      <Pagination requestApi={requestApi} next={pagination.next} prev={pagination.prev} />
    </div>
  }
}
