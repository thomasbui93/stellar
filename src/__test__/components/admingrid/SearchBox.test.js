import React from 'react'
import renderer from 'react-test-renderer'
import {SearchBox} from '../../../components/admingrid/SearchBox'
import {expect as expectChai} from 'chai'
import {spy} from 'sinon'
import {mount} from 'enzyme'

describe('SearchBox', () => {
  it('render correctly without error', () => {
    const dummyApi = () => {
      console.log('dummy check')
    }
    const component = renderer.create(
      <SearchBox requestApi={dummyApi} className='search-box-container' />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('trigger requestApi when submit the form with more than 2 character input (TBD)', () => {
    const dummyApi = spy()
    const component = mount(<SearchBox requestApi={dummyApi} />)

    component.find('button[type="button"]')
      .simulate('click')
    expectChai(dummyApi.called).to.equal(true)
    expectChai(dummyApi.calledWith({query: false}))
  })
})
