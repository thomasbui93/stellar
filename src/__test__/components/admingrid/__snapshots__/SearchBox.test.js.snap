// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox render correctly without error 1`] = `
<form
  className="search-box search-box-container"
  onSubmit={[Function]}
>
  <input
    placeholder="Enter your search here"
    type="text"
  />
  <button
    type="submit"
  >
    Search
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Clear
  </button>
</form>
`;
